### PR TITLE
dl tutorial files to tmp directory, then move them once successful

### LINKED ISSUE #1392: tutorial datasets can be corrupted on dl with bad connection
Corrupted datasets then do not perform as expected during tutorial exercises. See the list discussion for more details: https://groups.google.com/forum/#!topic/xarray/F5UKkZZsMec

### PR DESCRIPTION
**Pull request by gidden**
closes #1392

 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

I'm not sure how to best add tests for this. Also not sure how best to import `shutil`. I followed the current pattern for `os`, but am not sure why that pattern exists in this file. I can clean up further if this is deemed useful.

### CODE REVIEW

File: xarray/tutorial.py [hunk 1]
```diff
@@ -42,18 +43,27 @@ def load_dataset(name, cache=True, cache_dir=_default_cache_dir,
...
+            _os.mkdir(tmpdir)
 
         url = '/'.join((github_url, 'raw', 'master', fullname))
-        _urlretrieve(url, localfile)
+        _urlretrieve(url, tmpfile)
+
+        if not _os.path.exists(tmpfile):
+            raise ValueError('File could not be downloaded, please try again')
+
+        _shutil.move(tmpfile, localfile)
```
Comment (fmaussion) [inline xarray/tutorial.py]: I thought that the issue wasn't that the file isn't downloaded, but rather that it is incompletely downloaded? Will this new temporary step than solve this? 

Is there a way to make ``urlretrieve`` more robust, with checksums or something?

**Comment by gidden** [comment]
It seems like urlretrieve is coming directly from urllib. It seems in both
Python 2 and 3 ContentTooShort error should be thrown and isnt. Perhaps
checksums are the right way to go.

**Comment by gidden** [comment]
Successful locally with pydata/xarray-data#9:

```
In [3]: xr.tutorial.load_dataset('air_temperature', github_url='https://github.com/gidden/xarray-data', branch='md5')
Out[3]: 
<xarray.Dataset>
Dimensions:  (lat: 25, lon: 53, time: 2920)
Coordinates:
  * lat      (lat) float32 75.0 72.5 70.0 67.5 65.0 62.5 60.0 57.5 55.0 52.5 ...
  * lon      (lon) float32 200.0 202.5 205.0 207.5 210.0 212.5 215.0 217.5 ...
  * time     (time) datetime64[ns] 2013-01-01 2013-01-01T06:00:00 ...
Data variables:
    air      (time, lat, lon) float64 241.2 242.5 243.5 244.0 244.1 243.9 ...
Attributes:
    Conventions:  COARDS
    title:        4x daily NMC reanalysis (1948)
    description:  Data is from NMC initialized reanalysis\n(4x/day).  These a...
    platform:     Model
    references:   http://www.esrl.noaa.gov/psd/data/gridded/data.ncep.reanaly...
```

**Comment by shoyer** [comment]
Thanks for taking this on @gidden.

> Also not sure how best to import shutil. I followed the current pattern for os, but am not sure why that pattern exists in this file.

I think this exists to work around the fact that IPython shows everything that isn't private in auto-complete (https://github.com/ipython/ipykernel/issues/129). But we should probably switch back to normal imports in this module -- this is really more of an IPython issue.

**Comment by gidden** [comment]
Ok, sounds good. `shutil` was no longer needed, but the md5 checksum was
implemented as private. So I think all is well.

File: xarray/tutorial.py [hunk 2]
```diff
@@ -18,9 +20,17 @@
 _default_cache_dir = _os.sep.join(('~', '.xarray_tutorial_data'))
 
 
+def _md5(fname):
```
Comment (shoyer) [inline xarray/tutorial.py]: Call this something more descriptive like `check_file_md5_hash`?

Comment (gidden) [inline xarray/tutorial.py]: should it remain private (for ipython completion) or no?

Comment (shoyer) [inline xarray/tutorial.py]: I wouldn't stress about that. Given that we use it in the script in xarray-data, it's fine either way for me.

**Comment by gidden** [comment]
ok, this should be good to go

**Comment by shoyer** [comment]
Thanks @gidden, this look great to me. Could you kindly add a brief note to the "Bug fixes" section of `whats-new.rst`?

**Comment by shoyer** [comment]
I'd be slightly more comfortable if we had this being unit tested on Travis-CI, but that's somewhat tricky to setup because we don't wan to require network access for the test suite by default. You could do this by adding a flag like `--run-network-tests` in `conftest.py`, and then hooking it through in `xarray/tests/__init__.py` like `--run-flaky`. I would be awesome if you could take a look at that and set the Travis/Appveyor tests to test the tutorial, but isn't strictly required to merge this.

**Comment by gidden** [comment]
Hey @shoyer, if I have time later I will try to address the unit test issue. If I'm too late, feel free to pull this in the meantime.

**Comment by gidden** [comment]
hey @shoyer, @fmaussion. test decorator added with tutorial dataset unit test. i think this is good to go.

File: xarray/tests/test_tutorial.py [hunk 3]
```diff
@@ -1,23 +1,26 @@
...
+class TestLoadDataset(TestCase):
 
     def setUp(self):
         self.testfile = 'tiny'
         self.testfilepath = os.path.expanduser(os.sep.join(
             ('~', '.xarray_tutorial_data', self.testfile)))
         with suppress(OSError):
-            os.remove(self.testfilepath)
+            os.remove('{}.nc'.format(self.testfilepath))
+            os.remove('{}.md5'.format(self.testfilepath))
```
Comment (shoyer) [inline xarray/tests/test_tutorial.py]: This should go in its own `suppress` block

Comment (gidden) [inline xarray/tests/test_tutorial.py]: done!

**Comment by shoyer** [comment]
thanks!

**Pull request merged**